presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java

```diff
@@ -698,82 +701,44 @@ public static double round(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(B
         return Math.round(num * factor) / factor;
     }
 
-    private static SqlScalarFunction decimalRoundFunction()
-    {
-        Signature signature = Signature.builder()
-                .kind(SCALAR)
-                .name("round")
-                .longVariableConstraints(
-                        longVariableExpression("return_precision", "min(38, num_precision - num_scale + min(1, num_scale))"))
-                .argumentTypes(parseTypeSignature("decimal(num_precision,num_scale)", ImmutableSet.of("num_precision", "num_scale")))
-                .returnType(parseTypeSignature("decimal(return_precision,0)", ImmutableSet.of("return_precision")))
-                .build();
-        return SqlScalarFunction.builder(MathFunctions.class)
-                .signature(signature)
-                .description("round to nearest integer")
-                .implementation(b -> b
-                        .methods("roundShortDecimal")
-                        .withExtraParameters(MathFunctions::decimalRoundShortExtraParameters)
-                )
-                .implementation(b -> b
-                        .methods("roundLongDecimal")
-                        .withExtraParameters(MathFunctions::decimalRoundLongExtraParameters)
-                )
-                .build();
-    }
-
-    private static List<Object> decimalRoundShortExtraParameters(SpecializeContext context)
+    @ScalarFunction("round")
+    @Description("round to nearest integer")
+    public static final class Round
     {
-        long scale = context.getLiteral("num_scale");
-        long rescaleFactor = longTenToNth((int) scale);
-        return ImmutableList.of(rescaleFactor, scale);
-    }
+        @LiteralParameters({"p", "s", "rp", "rs"})
+        @SqlType("decimal(rp, rs)")
+        @Constraint(variable = "rp", expression = "min(38, p - s + min(1, s))")
+        @Constraint(variable = "rs", expression = "0")
+        public static long roundShort(@LiteralParameter("s") Long numScale, @SqlType("decimal(p, s)") long num)
```

Comment:
Now `@LiteralParameter` work only with `Long` - not with `long`

`num` is OK. Scalar definitions for other types use `num`.